src/commands/coverage/upload.ts

```diff
@@ -0,0 +1,356 @@
+        if (isFile(basePath)) {
+          return acc.concat(fs.existsSync(basePath) ? [basePath] : [])
+        }
+        let globPattern
+        // It's either a folder or a glob pattern
+        if (glob.hasMagic(basePath)) {
+          // It's a glob pattern so we just use it as is
+          globPattern = basePath
+        } else {
+          // It's a folder
+          globPattern = buildPath(basePath, '*.xml') // TODO update the logic to not assume that every report has .xml extension
+        }
+
+        const filesToUpload = glob.sync(globPattern).filter((file) => path.extname(file) === '.xml')
+
+        return acc.concat(filesToUpload)
+      }, [])
+      .filter(isFile)
+
+    const uniqueFiles = [...new Set(codeCoverageReportFiles)]
+    const pathsByFormat: {format: string; paths: string[]}[] = []
+
+    for (const codeCoverageReportPath of uniqueFiles) {
+      const format: string | undefined = (await detectFormat(codeCoverageReportPath)) || this.format
+      const validationErrorMessage = await validateCoverageReport(codeCoverageReportPath, format, this.format)
+      if (validationErrorMessage) {
+        this.context.stdout.write(renderInvalidFile(codeCoverageReportPath, validationErrorMessage))
+      } else {
+        let formatEntry = pathsByFormat.find((entry) => entry.format === format)
+        if (!formatEntry) {
+          formatEntry = {format: format as string, paths: []}
+          pathsByFormat.push(formatEntry)
+        }
+        formatEntry.paths.push(codeCoverageReportPath)
+      }
+    }
+
+    return pathsByFormat
+  }
+
+  private async uploadCodeCoverageReport(api: APIHelper, codeCoverageReport: Payload) {
+    if (this.dryRun) {
+      this.logger.info(renderDryRunUpload(codeCoverageReport))
+
+      return
+    }
+
+    try {
+      this.logger.info(renderUpload(codeCoverageReport))
+      await retryRequest(() => api.uploadCodeCoverageReport(codeCoverageReport), {
+        onRetry: (e, attempt) => {
+          this.context.stderr.write(renderRetriedUpload(codeCoverageReport, e.message, attempt))
+        },
+        retries: 5,
+      })
+    } catch (error) {
+      this.context.stderr.write(renderFailedUpload(codeCoverageReport, error))
+      if (error.response) {
+        // If it's an axios error
+        if (!errorCodesStopUpload.includes(error.response.status)) {
+          // And a status code that should not stop the whole upload, just return
+          return
+        }
+      }
+      throw error
+    }
+  }
+
+  private async uploadToGitDB(opts: {requestBuilder: RequestBuilder}) {
+    await uploadToGitDB(this.logger, opts.requestBuilder, await newSimpleGit(), this.dryRun, this.gitRepositoryURL)
+  }
+
+  private async uploadGitMetadata() {
+    if (await isGitRepo()) {
+      const traceId = id()
+
+      const requestBuilder = getRequestBuilder({
+        baseUrl: apiUrl,
+        apiKey: this.config.apiKey!,
+        headers: new Map([
+          [TRACE_ID_HTTP_HEADER, traceId],
```

Comment:
are we going to use distributed tracing for the request? Or is it here just because it's for Junit xml upload?

Reply:
I included it because it was in JUnit XML, is it not going to help when debugging things in the backend?

Reply:
yeah! I was just curious whether it was conscious or more copy-pasty 😄 . Let's leave it